src/core/model/table/dynamic_position_list_index.cpp

```diff
@@ -0,0 +1,172 @@
+#include "model/table/dynamic_position_list_index.h"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <map>
+#include <memory>
+#include <stack>
+#include <utility>
+
+#include <boost/dynamic_bitset.hpp>
+#include <easylogging++.h>
+
+namespace {
+// O(cluster)
+void ProcessDeleteToCluster(model::DynamicPositionListIndex::Cluster& cluster,
+                            std::unordered_set<size_t>& deleted_records_ids) {
+    auto records_pred = [&](size_t element) {
+        auto it = deleted_records_ids.find(element);
+        if (it != deleted_records_ids.end()) {
+            deleted_records_ids.erase(it);
+            return true;
+        }
+        return false;
+    };
+    std::erase_if(cluster, records_pred);
+}
+
+}  // namespace
+
+namespace model {
+
+DynamicPositionListIndex::DynamicPositionListIndex(ClusterCollection clusters, size_t size,
+                                                   size_t relation_size)
+    : clusters_(std::move(clusters)),
+      valid_records_number_(size),
+      next_record_id_(relation_size),
+      probing_table_cache_() {
+    ForceCacheProbingTable();
+}
+
+std::unique_ptr<DynamicPositionListIndex> DynamicPositionListIndex::CreateFor(
+        std::vector<ClusterValue>& records) {
+    ClusterCollection clusters;
+    for (size_t record_id = 0; record_id < records.size(); ++record_id) {
+        ClusterValue const& value = records[record_id];
+        clusters[value].push_back(record_id);
+    }
+
+    return std::make_unique<DynamicPositionListIndex>(std::move(clusters), records.size(),
+                                                      records.size());
+}
+
+void DynamicPositionListIndex::UpdateWith(
+        std::vector<std::pair<std::optional<size_t>, ClusterValue>>& inserted_records,
+        std::unordered_set<size_t> deleted_records_ids) {
+    if (inserted_records.empty() && deleted_records_ids.empty()) {
+        return;
+    }
+    for (auto it = clusters_.begin(); it != clusters_.end();) {
+        auto& [cluster_value, cluster] = *it;
+        ProcessDeleteToCluster(cluster, deleted_records_ids);
+        if (cluster.empty()) {
+            it = clusters_.erase(it);
+        } else {
+            ++it;
+        }
+    }
+    for (auto const& [encoded_value_opt, cluster_value] : inserted_records) {
+        if (encoded_value_opt.has_value()) {
+            clusters_[cluster_value].emplace_back(*encoded_value_opt);
+        } else {
+            clusters_[cluster_value].emplace_back(next_record_id_++);
+            ++valid_records_number_;
+        }
+    }
+    valid_records_number_ -= deleted_records_ids.size();
```

Comment:
Save `deleted_records_ids.size()` into a variable before the deletion loop, this always subtracts zero.